File: codes/matlab/ch17/stability_regions.m
% stability_regions.m
% Chapter 17: Stability of Spectral Methods
%
% Computational Etude 17.2: Stability Regions and Spectral Eigenvalue Overlay.
%
% Two figures:
%   (1) 2x3 grid showing stability regions of six classical time-stepping
%       methods: Forward Euler, RK4, Leapfrog, Adams-Bashforth 2,
%       Crank-Nicolson, Backward Euler.
%   (2) 1x2 figure overlaying spectral eigenvalues on stability regions
%       of RK4 (Fourier) and Forward Euler (Chebyshev D^2).
%
% Author: Dr. Denys Dutykh
% Part of "Computational Etudes: A Spectral Approach"

clear; clc; close all;

NAVY       = [20, 45, 110]/255;
SKY        = [120, 150, 210]/255;
CORAL      = [231, 76, 60]/255;
TEAL       = [22, 160, 133]/255;
AMBER      = [230, 126, 34]/255;
LIGHT_NAVY = [197, 208, 232]/255;

%% ---- Figure 1: Stability regions (2x3 grid) ----
figure('Position', [50, 50, 1200, 800]);

% (a) Forward Euler
subplot(2, 3, 1);
plot_stability_contour(@(z) 1 + z, [-3, 1.5], [-2, 2], ...
    '(a) Forward Euler', NAVY, LIGHT_NAVY);

% (b) RK4
subplot(2, 3, 2);
plot_stability_contour(@(z) 1 + z + z.^2/2 + z.^3/6 + z.^4/24, ...
    [-5, 2], [-4, 4], '(b) RK4', NAVY, LIGHT_NAVY);

% (c) Leapfrog -- parametric boundary
subplot(2, 3, 3);
theta = linspace(0, 2*pi, 1000);
g = exp(1i*theta);
z_lf = (g.^2 - 1) ./ (2*g);
fill(real(z_lf), imag(z_lf), LIGHT_NAVY, 'FaceAlpha', 0.5, ...
     'EdgeColor', NAVY, 'LineWidth', 1.5);
hold on;
plot([min(real(z_lf))-0.5, max(real(z_lf))+0.5], [0, 0], ...
     'Color', [0.5 0.5 0.5], 'LineWidth', 0.5);
plot([0, 0], [-1.5, 1.5], 'Color', [0.5 0.5 0.5], 'LineWidth', 0.5);
hold off;
xlim([-1.5, 1.5]); ylim([-1.5, 1.5]);
xlabel('Re(z)'); ylabel('Im(z)');
title('(c) Leapfrog');
axis equal; grid on; set(gca, 'GridAlpha', 0.2);

% (d) Adams-Bashforth 2 -- parametric boundary
subplot(2, 3, 4);
z_ab = (g.^2 - g) ./ (1.5*g - 0.5);
fill(real(z_ab), imag(z_ab), LIGHT_NAVY, 'FaceAlpha', 0.5, ...
     'EdgeColor', NAVY, 'LineWidth', 1.5);
hold on;
plot([-2, 1.5], [0, 0], 'Color', [0.5 0.5 0.5], 'LineWidth', 0.5);
plot([0, 0], [-1.5, 1.5], 'Color', [0.5 0.5 0.5], 'LineWidth', 0.5);
hold off;
xlim([-2, 1.5]); ylim([-1.5, 1.5]);
xlabel('Re(z)'); ylabel('Im(z)');
title('(d) Adams-Bashforth 2');
axis equal; grid on; set(gca, 'GridAlpha', 0.2);

% (e) Crank-Nicolson
subplot(2, 3, 5);
plot_stability_contour(@(z) (1 + z/2)./(1 - z/2), ...
    [-5, 1], [-4, 4], '(e) Crank-Nicolson', NAVY, LIGHT_NAVY);

% (f) Backward Euler
subplot(2, 3, 6);
plot_stability_contour(@(z) 1./(1 - z), ...
    [-1.5, 5], [-4, 4], '(f) Backward Euler', NAVY, LIGHT_NAVY);

set(gcf, 'PaperPositionMode', 'auto');
print('-dpdf', 'stability_regions.pdf');
fprintf('Figure saved to stability_regions.pdf\n');

%% ---- Figure 2: Spectra overlay ----
figure('Position', [100, 100, 1100, 500]);

% Left panel: RK4 + Fourier imaginary eigenvalues
subplot(1, 2, 1);
n_grid = 400;
xv = linspace(-5, 2, n_grid);
yv = linspace(-4, 4, n_grid);
[X, Y] = meshgrid(xv, yv);
Z = X + 1i*Y;
G = abs(1 + Z + Z.^2/2 + Z.^3/6 + Z.^4/24);
contourf(X, Y, G, [0, 1], 'FaceColor', LIGHT_NAVY, 'FaceAlpha', 0.4);
hold on;
contour(X, Y, G, [1, 1], 'LineColor', NAVY, 'LineWidth', 1.5);

% Fourier eigenvalues for advection
N_f = 32;
kf = [0:N_f/2-1, -N_f/2:-1];
lam_f = 1i * kf;

% Stable dt
dt_s = 2.8 / (N_f/2);
eigs_s = dt_s * lam_f;
plot(real(eigs_s), imag(eigs_s), 'o', 'Color', TEAL, 'MarkerSize', 4, ...
     'MarkerFaceColor', TEAL, ...
     'DisplayName', sprintf('dt = %.4f (stable)', dt_s));

% Unstable dt
dt_u = 3.5 / (N_f/2);
eigs_u = dt_u * lam_f;
plot(real(eigs_u), imag(eigs_u), 's', 'Color', CORAL, 'MarkerSize', 4, ...
     'MarkerFaceColor', CORAL, ...
     'DisplayName', sprintf('dt = %.4f (unstable)', dt_u));

plot([-5, 2], [0, 0], 'Color', [0.5 0.5 0.5], 'LineWidth', 0.5, ...
     'HandleVisibility', 'off');
plot([0, 0], [-4, 4], 'Color', [0.5 0.5 0.5], 'LineWidth', 0.5, ...
     'HandleVisibility', 'off');
hold off;
xlabel('Re(\Delta t \lambda)'); ylabel('Im(\Delta t \lambda)');
title(sprintf('(a) RK4 + Fourier advection (N=%d)', N_f));
axis equal; legend('Location', 'southwest', 'FontSize', 8);
grid on; set(gca, 'GridAlpha', 0.2);

% Right panel: Forward Euler + Chebyshev D^2 eigenvalues
subplot(1, 2, 2);
xv2 = linspace(-3, 1.5, n_grid);
yv2 = linspace(-2, 2, n_grid);
[X2, Y2] = meshgrid(xv2, yv2);
Z2 = X2 + 1i*Y2;
G2 = abs(1 + Z2);
contourf(X2, Y2, G2, [0, 1], 'FaceColor', LIGHT_NAVY, 'FaceAlpha', 0.4);
hold on;
contour(X2, Y2, G2, [1, 1], 'LineColor', NAVY, 'LineWidth', 1.5);

% Chebyshev D^2 eigenvalues (Dirichlet)
N_c = 16;
[D, ~] = cheb(N_c);
D2 = D * D;
D2_int = D2(2:N_c, 2:N_c);
eigs_c = eig(D2_int);

% Scale dt so a few eigenvalues fit in the disk
phys_mask = abs(real(eigs_c)) < 200;
out_mask  = ~phys_mask;
dt_c = 0.8 / max(abs(real(eigs_c(1:5))));
scaled_eigs = dt_c * eigs_c;

plot(real(scaled_eigs(phys_mask)), imag(scaled_eigs(phys_mask)), 'o', ...
     'Color', TEAL, 'MarkerSize', 5, 'MarkerFaceColor', TEAL, ...
     'DisplayName', 'physical modes');
if any(out_mask)
    plot(real(scaled_eigs(out_mask)), imag(scaled_eigs(out_mask)), 's', ...
         'Color', CORAL, 'MarkerSize', 5, 'MarkerFaceColor', CORAL, ...
         'DisplayName', 'outlier modes');
end

plot([-3, 1.5], [0, 0], 'Color', [0.5 0.5 0.5], 'LineWidth', 0.5, ...
     'HandleVisibility', 'off');
plot([0, 0], [-2, 2], 'Color', [0.5 0.5 0.5], 'LineWidth', 0.5, ...
     'HandleVisibility', 'off');
hold off;
xlabel('Re(\Delta t \lambda)'); ylabel('Im(\Delta t \lambda)');
title(sprintf('(b) Forward Euler + Chebyshev D^2 (N=%d)', N_c));
axis equal; legend('Location', 'northwest', 'FontSize', 8);
grid on; set(gca, 'GridAlpha', 0.2);

set(gcf, 'PaperPositionMode', 'auto');
print('-dpdf', 'spectra_overlay.pdf');
fprintf('Figure saved to spectra_overlay.pdf\n');


%% ---- Helper function: contour-based stability region ----
function plot_stability_contour(g_fun, xlims, ylims, ttl, navy, light_navy)
    n_grid = 400;
    xv = linspace(xlims(1), xlims(2), n_grid);
    yv = linspace(ylims(1), ylims(2), n_grid);
    [X, Y] = meshgrid(xv, yv);
    Z = X + 1i*Y;
    G = abs(g_fun(Z));
    contourf(X, Y, G, [0, 1], 'FaceColor', light_navy, 'FaceAlpha', 0.5);
    hold on;
    contour(X, Y, G, [1, 1], 'LineColor', navy, 'LineWidth', 1.5);
    plot([xlims(1), xlims(2)], [0, 0], 'Color', [0.5 0.5 0.5], ...
         'LineWidth', 0.5);
    plot([0, 0], [ylims(1), ylims(2)], 'Color', [0.5 0.5 0.5], ...
         'LineWidth', 0.5);
    hold off;
    xlim(xlims); ylim(ylims);
    xlabel('Re(z)'); ylabel('Im(z)');
    title(ttl);
    axis equal; grid on; set(gca, 'GridAlpha', 0.2);
end

File: codes/julia/ch17/stability_regions.jl
#!/usr/bin/env julia
#=
stability_regions.jl
Chapter 17: Stability of Spectral Methods

Computational Etude 17.2: Stability Regions and Spectral Eigenvalue Overlay.

Two figures:
  (1) stability_regions.pdf -- 2x3 grid showing stability regions of six
      classical time-stepping methods: Forward Euler, RK4, Leapfrog,
      Adams-Bashforth 2, Crank-Nicolson, Backward Euler.
  (2) spectra_overlay.pdf -- 1x2 figure overlaying spectral eigenvalues
      on the stability regions of RK4 and Forward Euler.

Generates Figures 17.2a,b: stability_regions.pdf, spectra_overlay.pdf

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
=#

using LinearAlgebra
using Plots
using LaTeXStrings

gr()

const NAVY       = RGB(20/255, 45/255, 110/255)
const SKY        = RGB(120/255, 150/255, 210/255)
const CORAL      = RGB(231/255, 76/255, 60/255)
const TEAL       = RGB(22/255, 160/255, 133/255)
const AMBER      = RGB(230/255, 126/255, 34/255)
const LIGHT_NAVY = RGB(197/255, 208/255, 232/255)

const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch17", "julia")
mkpath(OUTPUT_DIR)

# --- Stability functions ---
g_fe(z)  = 1.0 .+ z
g_rk4(z) = 1.0 .+ z .+ z.^2 ./2 .+ z.^3 ./6 .+ z.^4 ./24
g_be(z)  = 1.0 ./ (1.0 .- z)
g_cn(z)  = (1.0 .+ z./2) ./ (1.0 .- z./2)

"""Boundary of AB2 stability region via root locus on |g|=1."""
function ab2_boundary(n_pts=500)
    theta = range(0, 2pi, length=n_pts)
    g = exp.(1im .* theta)
    z = (g.^2 .- g) ./ (1.5 .* g .- 0.5)
    return real.(z), imag.(z)
end

"""Boundary of leapfrog stability region via root locus on |g|=1."""
function leapfrog_boundary(n_pts=500)
    theta = range(0, 2pi, length=n_pts)
    g = exp.(1im .* theta)
    z = (g.^2 .- 1.0) ./ (2.0 .* g)
    return real.(z), imag.(z)
end

"""Chebyshev differentiation matrix."""
function cheb(N)
    if N == 0; return [0.0;;], [1.0]; end
    x = [cos(pi * j / N) for j in 0:N]
    c = [j == 0 || j == N ? 2.0 : 1.0 for j in 0:N] .* [(-1.0)^j for j in 0:N]
    X = repeat(x, 1, N+1)
    dX = X - X' + I(N+1)
    D = (c * (1.0 ./ c)') ./ dX
    D .-= Diagonal(vec(sum(D, dims=2)))
    return D, x
end

"""Plot stability region for one-step methods using contour fill."""
function plot_stability_contour!(p, gfun, xlims, ylims, title_str; n_grid=400)
    xr = range(xlims[1], xlims[2], length=n_grid)
    yr = range(ylims[1], ylims[2], length=n_grid)
    Z = [complex(xi, yi) for yi in yr, xi in xr]
    G = abs.(gfun(Z))

    # Shade stable region
    contourf!(p, xr, yr, G, levels=[0, 1], c=cgrad([LIGHT_NAVY, LIGHT_NAVY]),
              alpha=0.5, colorbar=false)
    contour!(p, xr, yr, G, levels=[1], c=NAVY, lw=1.5, colorbar=false)
    hline!(p, [0], color=:gray, lw=0.5, label="")
    vline!(p, [0], color=:gray, lw=0.5, label="")
    plot!(p, xlabel=L"\mathrm{Re}(z)", ylabel=L"\mathrm{Im}(z)",
          title=title_str, aspect_ratio=:equal,
          xlims=xlims, ylims=ylims,
          grid=true, gridalpha=0.2)
end

"""Plot stability region for multistep methods using fill."""
function plot_stability_boundary!(p, bx, by, xlims, ylims, title_str)
    plot!(p, bx, by, seriestype=:shape, fillcolor=LIGHT_NAVY,
          fillalpha=0.5, linecolor=NAVY, lw=1.5, label="")
    hline!(p, [0], color=:gray, lw=0.5, label="")
    vline!(p, [0], color=:gray, lw=0.5, label="")
    plot!(p, xlabel=L"\mathrm{Re}(z)", ylabel=L"\mathrm{Im}(z)",
          title=title_str, aspect_ratio=:equal,
          xlims=xlims, ylims=ylims,
          grid=true, gridalpha=0.2)
end

function figure_stability_regions()
    # Create 6 subplots
    p1 = plot()
    plot_stability_contour!(p1, g_fe, (-3, 1.5), (-2, 2), "(a) Forward Euler")

    p2 = plot()
    plot_stability_contour!(p2, g_rk4, (-5, 2), (-4, 4), "(b) RK4")

    p3 = plot()
    bx_lf, by_lf = leapfrog_boundary(1000)
    plot_stability_boundary!(p3, bx_lf, by_lf, (-1.5, 1.5), (-1.5, 1.5), "(c) Leapfrog")

    p4 = plot()
    bx_ab, by_ab = ab2_boundary(1000)
    plot_stability_boundary!(p4, bx_ab, by_ab, (-2, 1.5), (-1.5, 1.5), "(d) Adams-Bashforth 2")

    p5 = plot()
    plot_stability_contour!(p5, g_cn, (-5, 1), (-4, 4), "(e) Crank-Nicolson")

    p6 = plot()
    plot_stability_contour!(p6, g_be, (-1.5, 5), (-4, 4), "(f) Backward Euler")

    fig = plot(p1, p2, p3, p4, p5, p6, layout=(2, 3), size=(1200, 800))

    savefig(fig, joinpath(OUTPUT_DIR, "stability_regions.pdf"))
    savefig(fig, joinpath(OUTPUT_DIR, "stability_regions.png"))
    println("Figure saved to $(joinpath(OUTPUT_DIR, "stability_regions.pdf"))")
    return fig
end

function figure_spectra_overlay()
    n_grid = 400

    # --- Left panel: RK4 + Fourier imaginary eigenvalues ---
    xr1 = range(-5, 2, length=n_grid)
    yr1 = range(-4, 4, length=n_grid)
    Z1 = [complex(xi, yi) for yi in yr1, xi in xr1]
    G1 = abs.(g_rk4(Z1))

    p1 = plot()
    contourf!(p1, xr1, yr1, G1, levels=[0, 1], c=cgrad([LIGHT_NAVY, LIGHT_NAVY]),
              alpha=0.4, colorbar=false)
    contour!(p1, xr1, yr1, G1, levels=[1], c=NAVY, lw=1.5, colorbar=false)

    # Fourier eigenvalues for advection: lambda_k = i*k
    N_f = 32
    kvals = collect(-N_f÷2+1 : N_f÷2)
    lam_f = 1im .* kvals

    # Stable dt
    dt_s = 2.8 / (N_f / 2)
    eigs_s = dt_s .* lam_f
    scatter!(p1, real.(eigs_s), imag.(eigs_s), color=TEAL, ms=3,
             markershape=:circle, label=latexstring("\\Delta t = $(round(dt_s, digits=4)) (stable)"))

    # Unstable dt
    dt_u = 3.5 / (N_f / 2)
    eigs_u = dt_u .* lam_f
    scatter!(p1, real.(eigs_u), imag.(eigs_u), color=CORAL, ms=3,
             markershape=:square, label=latexstring("\\Delta t = $(round(dt_u, digits=4)) (unstable)"))

    hline!(p1, [0], color=:gray, lw=0.5, label="")
    vline!(p1, [0], color=:gray, lw=0.5, label="")
    plot!(p1, xlabel=L"\mathrm{Re}(\Delta t \lambda)", ylabel=L"\mathrm{Im}(\Delta t \lambda)",
          title=L"(a) RK4 + Fourier advection ($N=32$)",
          aspect_ratio=:equal, grid=true, gridalpha=0.2, legendfontsize=7)

    # --- Right panel: Forward Euler + Chebyshev D^2 eigenvalues ---
    xr2 = range(-3, 1.5, length=n_grid)
    yr2 = range(-2, 2, length=n_grid)
    Z2 = [complex(xi, yi) for yi in yr2, xi in xr2]
    G2 = abs.(g_fe(Z2))

    p2 = plot()
    contourf!(p2, xr2, yr2, G2, levels=[0, 1], c=cgrad([LIGHT_NAVY, LIGHT_NAVY]),
              alpha=0.4, colorbar=false)
    contour!(p2, xr2, yr2, G2, levels=[1], c=NAVY, lw=1.5, colorbar=false)

    # Chebyshev D^2 eigenvalues (Dirichlet)
    N_c = 16
    D, xc = cheb(N_c)
    D2 = D * D
    D2_int = D2[2:N_c, 2:N_c]
    eigs_c = eigvals(D2_int)

    # Scale: choose dt so a few eigenvalues fit in the disk
    sorted_eigs = sort(abs.(real.(eigs_c)))
    dt_c = 0.8 / sorted_eigs[min(5, length(sorted_eigs))]
    scaled_eigs = dt_c .* eigs_c

    # Separate physical from outlier modes
    phys_mask = abs.(real.(eigs_c)) .< 200
    out_mask  = .!phys_mask

    scatter!(p2, real.(scaled_eigs[phys_mask]), imag.(scaled_eigs[phys_mask]),
             color=TEAL, ms=4, markershape=:circle, label="physical modes")
    if any(out_mask)
        scatter!(p2, real.(scaled_eigs[out_mask]), imag.(scaled_eigs[out_mask]),
                 color=CORAL, ms=4, markershape=:square, label="outlier modes")
    end

    hline!(p2, [0], color=:gray, lw=0.5, label="")
    vline!(p2, [0], color=:gray, lw=0.5, label="")
    plot!(p2, xlabel=L"\mathrm{Re}(\Delta t \lambda)", ylabel=L"\mathrm{Im}(\Delta t \lambda)",
          title=L"(b) Fwd Euler + Chebyshev $D^2$ ($N=16$)",
          aspect_ratio=:equal, grid=true, gridalpha=0.2, legendfontsize=7)

    fig = plot(p1, p2, layout=(1, 2), size=(1100, 500))

    savefig(fig, joinpath(OUTPUT_DIR, "spectra_overlay.pdf"))
    savefig(fig, joinpath(OUTPUT_DIR, "spectra_overlay.png"))
    println("Figure saved to $(joinpath(OUTPUT_DIR, "spectra_overlay.pdf"))")
    return fig
end

function main()
    figure_stability_regions()
    figure_spectra_overlay()
end

main()
